proof/refine/ARM/SchedContextInv_R.thy

```diff
@@ -70,14 +70,65 @@ primrec sc_ctrl_inv_rel ::
   "sc_ctrl_inv_rel (Invocations_A.InvokeSchedControlConfigure sc budget period refills badge) sci' =
     (sci' = InvokeSchedControlConfigure sc budget period refills badge)"
 
-(* FIXME RT: preconditions can be reduced, this is what is available at the call site: *)
+lemma decodeSchedContext_Bind_wf:
+  "\<lbrace>valid_cap' (SchedContextCap sc n) and ex_nonz_cap_to' sc
+    and (\<lambda>s. \<forall>cap\<in>set excaps. \<forall>r\<in>zobj_refs' cap. ex_nonz_cap_to' r s)
+    and (\<lambda>s. \<forall>x\<in>set excaps. valid_cap' x s)\<rbrace>
+   decodeSchedContext_Bind sc excaps
+   \<lbrace>valid_sc_inv'\<rbrace>, -"
+  apply (clarsimp simp: decodeSchedContext_Bind_def)
+  apply (wpsimp wp: gts_wp' threadGet_wp getNotification_wp
+              simp: scReleased_def scActive_def isBlocked_def refillReady_def)
+  apply (clarsimp simp: valid_cap'_def)
+  apply (drule_tac x="hd excaps" in bspec, fastforce dest: hd_in_set)+
+  apply (fastforce simp: pred_tcb_at'_def obj_at'_def)
+  done
+
+lemma decodeSchedContext_UnbindObject_wf:
+  "\<lbrace>valid_cap' (SchedContextCap sc n) and ex_nonz_cap_to' sc
+    and (\<lambda>s. \<forall>cap\<in>set excaps. \<forall>r\<in>zobj_refs' cap. ex_nonz_cap_to' r s)
+    and (\<lambda>s. \<forall>x\<in>set excaps. valid_cap' x s)\<rbrace>
+   decodeSchedContext_UnbindObject sc excaps
+   \<lbrace>valid_sc_inv'\<rbrace>, -"
+  apply (clarsimp simp: decodeSchedContext_UnbindObject_def)
+  apply (wpsimp wp: gts_wp' threadGet_wp getNotification_wp
+              simp: scReleased_def scActive_def isBlocked_def refillReady_def)
+  apply (clarsimp simp: valid_cap'_def)
+  apply (drule_tac x="hd excaps" in bspec, fastforce dest: hd_in_set)+
+  apply (fastforce simp: pred_tcb_at'_def obj_at'_def)
+  done
+
+lemma decodeSchedContext_YieldTo_wf:
+  "\<lbrace>valid_cap' (SchedContextCap sc n) and ex_nonz_cap_to' sc\<rbrace>
+   decodeSchedContext_YieldTo sc args
+   \<lbrace>valid_sc_inv'\<rbrace>, -"
+  apply (clarsimp simp: decodeSchedContext_YieldTo_def)
+  apply (wpsimp wp: gts_wp' threadGet_wp getNotification_wp getTCB_wp
+              simp: scReleased_def scActive_def isBlocked_def refillReady_def)
+  apply (clarsimp simp: valid_cap'_def)
+  apply (clarsimp simp: pred_tcb_at'_def obj_at'_def projectKOs)
+  done
+
+lemma valid_SchedContextCap_sc_at':
```

Comment:
nitpick: perhaps move this to Invariants_H or somewhere else higher up from here.